cleverhans/future/tf2/attacks/carlini_wagner_l2.py

```diff
@@ -0,0 +1,334 @@
+"""The CarliniWagnerL2 attack.
+"""
+import numpy as np
+import tensorflow as tf
+
+from cleverhans.future.tf2.utils_tf import get_or_guess_labels, set_with_mask
+
+
+def carlini_wagner_l2(model_fn, x, **kwargs):
+  """
+  This is the function interface for the Carlini-Wagner-L2 attack.
+  For more details on the attack and the parameters see the corresponding class.
+  """
+  return CarliniWagnerL2(model_fn, **kwargs).attack(x)
+
+
+class CarliniWagnerL2(object):
+  def __init__(self,
+               model_fn,
+               y=None,
+               targeted=False,
+               batch_size=128,
+               clip_min=0.,
+               clip_max=1.,
+               binary_search_steps=5,
+               max_iterations=1_000,
+               abort_early=True,
+               confidence=0.,
+               initial_const=1e-2,
+               learning_rate=5e-3):
+    """
+    This attack was originally proposed by Carlini and Wagner. It is an
+    iterative attack that finds adversarial examples on many defenses that
+    are robust to other attacks.
+    Paper link: https://arxiv.org/abs/1608.04644
+    At a high level, this attack is an iterative attack using Adam and
+    a specially-chosen loss function to find adversarial examples with
+    lower distortion than other attacks. This comes at the cost of speed,
+    as this attack is often much slower than others.
+
+    :param model_fn: a callable that takes an input tensor and returns the model logits.
+    :param y: (optional) Tensor with target labels. 
+    :param y: (optional) Targeted attack? 
+    :param batch_size: Number of attacks to run simultaneously.
+    :param clip_min: (optional) float. Minimum float values for adversarial example components.
+    :param clip_max: (optional) float. Maximum float value for adversarial example components.
+    :param binary_search_steps: The number of times we perform binary
+                            search to find the optimal tradeoff-
+                            constant between norm of the purturbation
+                            and confidence of the classification.
+    :param max_iterations: The maximum number of iterations. Setting this
+                           to a larger value will produce lower distortion
+                           results. Using only a few iterations requires
+                           a larger learning rate, and will produce larger
+                           distortion results.
+    :param abort_early: If true, allows early aborts if gradient descent
+                    is unable to make progress (i.e., gets stuck in
+                    a local minimum).
+    :param confidence: Confidence of adversarial examples: higher produces
+                       examples with larger l2 distortion, but more
+                       strongly classified as adversarial.
+    :param initial_const: The initial tradeoff-constant to use to tune the
+                      relative importance of size of the pururbation
```

Comment:
Small typo ^^

Reply:
Most of those parameters are optional right? Can we mark them as such?

Reply:
Addressed in f414ab9